### PR TITLE
Limit use of ConditionalFact

### PR DESCRIPTION
**Pull request by jasonmalinowski**
We have a bunch of tests using ConditionalFact to limit where they run even though there's no reason they can't run in more places.

### CODE REVIEW

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/Compilers/Test/Core/ObjectReference.cs [hunk 1]
```diff
@@ -31,6 +31,25 @@ public static class ObjectReference
...
+    /// </code>
+    /// 
+    /// This code has a bug: it presumes that when strongReference = null is assigned, there are no other references anywhere.
+    /// But that line only tells the JIT to null out the place that's holding the active value. The JIT could have spilled a copy
+    /// at some point to the stack, which it now considers unused and isn't worth cleaning up. Or another register might still be
+    /// holding it, etc.
+    /// 
+    /// What this class does is it holds the only active reference in the heap, and any use of that reference is put in a method
+    /// that is marked NoInline; this ensures that when the uses are done, any temporaries still floating around are understood
+    /// by the JIT/GC to actually be unused.
```
Comment (sharwell) [inline src/Compilers/Test/Core/ObjectReference.cs:51]:
```suggestion
    /// <para>Specifically, consider this common pattern:</para>
    /// 
    /// <code>
    /// var weakReference = new WeakReference(strongReference);
    /// strongReference = null;
    /// GC.Collect(); // often a few times...
    /// Assert.Null(weakReference.Target);
    /// </code>
    /// 
    /// <para>This code has a bug: it presumes that when strongReference = null is assigned, there are no other references anywhere.
    /// But that line only tells the JIT to null out the place that's holding the active value. The JIT could have spilled a copy
    /// at some point to the stack, which it now considers unused and isn't worth cleaning up. Or another register might still be
    /// holding it, etc.</para>
    /// 
    /// <para>What this class does is it holds the only active reference in the heap, and any use of that reference is put in a method
    /// that is marked NoInline; this ensures that when the uses are done, any temporaries still floating around are understood
    /// by the JIT/GC to actually be unused.</para>
```

**Review by ryzngard — APPROVED** [review]

File: src/Compilers/Test/Core/ObjectReference.cs [hunk 2]
```diff
@@ -127,6 +155,18 @@ public U UseReference<U>(Func<T, U> function)
             return function(GetReferenceWithChecks());
         }
 
+        /// <summary>
+        /// Provides the underlying strong reference to the given function, lets a function extract some value, and then returns a new ObjectReference.
+        /// This method is marked not be inlined, to ensure that no temporaries are left on the stack that might still root the strong reference. The
+        /// caller must not "leak" the object out of the given action for any lifetime assertions to be safe.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        public ObjectReference<U> GetObjectReference<U>(Func<T, U> function) where U : class
```
Comment (sharwell) [inline src/Compilers/Test/Core/ObjectReference.cs:164]:
```suggestion
        public ObjectReference<U> Apply<U>(Func<T, U> function) where U : class
```

**Review by sharwell — APPROVED** [review]

**Review by jinujoseph — APPROVED** [review]

File: src/Compilers/Test/Core/ObjectReference.cs [hunk 3]
```diff
@@ -31,6 +31,25 @@ public static class ObjectReference
...
+    /// 
+    /// This code has a bug: it presumes that when strongReference = null is assigned, there are no other references anywhere.
+    /// But that line only tells the JIT to null out the place that's holding the active value. The JIT could have spilled a copy
+    /// at some point to the stack, which it now considers unused and isn't worth cleaning up. Or another register might still be
+    /// holding it, etc.
+    /// 
+    /// What this class does is it holds the only active reference in the heap, and any use of that reference is put in a method
+    /// that is marked NoInline; this ensures that when the uses are done, any temporaries still floating around are understood
+    /// by the JIT/GC to actually be unused.
+    /// </remarks>
```
Comment (jcouv) [inline src/Compilers/Test/Core/ObjectReference.cs:52]: nit: consider adding an example of correct/proper pattern to use instead.

**Review by jcouv — APPROVED** [review]
Compiler test file change LGTM Thanks (iteration 3)

**Review by jmarolf — APPROVED** [review]
:shipit:

**Comment by jmarolf** [comment]
>             var sol = ObjectReference.CreateFromFactory(() => MSBuildWorkspace.Create().OpenSolutionAsync(GetSolutionFileName("TestSolution.sln")).Result);

nice!

---
Refers to: src/Workspaces/MSBuildTest/MSBuildWorkspaceTests.cs:3032 in 83248ed. [](commit_id = 83248edef7f0ddd061a03deebabbf37592167b07, deletion_comment = False)

**Pull request merged**